src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/AssemblyBuilderImpl.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+
+namespace System.Reflection.Emit
+{
+    public sealed class AssemblyBuilderImpl : AssemblyBuilder
+    {
+        private bool _previouslySaved;
+        private AssemblyName _assemblyName;
+        private ModuleBuilderImpl? _module;
+
+        internal AssemblyBuilderImpl(AssemblyName name, IEnumerable<CustomAttributeBuilder>? assemblyAttributes)
+        {
+            ArgumentNullException.ThrowIfNull(name);
+
+            _assemblyName = name;
+
+            _module = new ModuleBuilderImpl(ModuleBuilderImpl.ManifestModuleName, this);
+
+            if (assemblyAttributes != null)
+            {
+                foreach (CustomAttributeBuilder assemblyAttribute in assemblyAttributes)
+                {
+                    SetCustomAttribute(assemblyAttribute);
+                }
+            }
+        }
+
+        public static AssemblyBuilderImpl DefineDynamicAssembly(AssemblyName name)
+            => new AssemblyBuilderImpl(name, null);
+
+        public static AssemblyBuilderImpl DefineDynamicAssembly(
+            AssemblyName name,
+            IEnumerable<CustomAttributeBuilder>? assemblyAttributes)
+                => new AssemblyBuilderImpl(name, assemblyAttributes);
+
+        private static void WritePEImage(Stream peStream, MetadataBuilder metadataBuilder, BlobBuilder ilBuilder) // MethodDefinitionHandle entryPointHandle when we have main method.
+        {
+            // Create executable with the managed metadata from the specified MetadataBuilder.
+            var peHeaderBuilder = new PEHeaderBuilder(
+                imageCharacteristics: Characteristics.Dll // Start off with a simple DLL
+                );
+
+            var peBuilder = new ManagedPEBuilder(
+                peHeaderBuilder,
+                new MetadataRootBuilder(metadataBuilder),
+                ilBuilder,
+                flags: CorFlags.ILOnly,
+                deterministicIdProvider: content => new BlobContentId(Guid.NewGuid(), 0x04030201)); // Const ID, will reexamine as project progresses.
+
+            // Write executable into the specified stream.
+            var peBlob = new BlobBuilder();
+            BlobContentId contentId = peBuilder.Serialize(peBlob);
+            peBlob.WriteContentTo(peStream);
+        }
+
+        public void Save(Stream stream)
+        {
+            if (_previouslySaved) // Cannot save an assembly multiple times. This is consistent with Save() in .Net Framework.
+            {
+                throw new InvalidOperationException(SR.CannotSaveMultipleTimes);
+            }
+
+            ArgumentNullException.ThrowIfNull(stream);
```

Comment:
```suggestion
            ArgumentNullException.ThrowIfNull(stream);

            if (_previouslySaved) // Cannot save an assembly multiple times. This is consistent with Save() in .Net Framework.
            {
                throw new InvalidOperationException(SR.CannotSaveMultipleTimes);
            }
```
Argument checks should be first.